internal/telemetry/metrics.go

```diff
@@ -0,0 +1,169 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"runtime"
+	"time"
+
+	"github.com/jzelinskie/cobrautil"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/rs/zerolog/log"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/authzed/spicedb/internal/datastore"
+	"github.com/authzed/spicedb/internal/middleware/usagemetrics"
+)
+
+var (
+	registry                  = prometheus.NewRegistry()
+	dynamicDispatchHistLabels = []string{"method", "cached"}
+)
+
+// RegisterTelemetryCollector registers a collector for the various pieces of
+// data required by SpiceDB telemetry.
+func RegisterTelemetryCollector(datastoreEngine string, ds datastore.Datastore) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	nodeID, err := os.Hostname()
+	if err != nil {
+		return fmt.Errorf("unable to get hostname: %w", err)
+	}
+
+	dbStats, err := ds.Statistics(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to query DB stats: %w", err)
+	}
+
+	clusterID := dbStats.UniqueID
+
+	if err := registry.Register(&collector{
+		ds: ds,
+		infoDesc: prometheus.NewDesc(
+			prometheus.BuildFQName("spicedb", "telemetry", "info"),
+			"Information about the SpiceDB environment.",
+			nil,
+			prometheus.Labels{
+				"cluster_id": clusterID,
+				"node_id":    nodeID,
+				"version":    cobrautil.Version,
+				"os":         runtime.GOOS,
+				"arch":       runtime.GOARCH,
+				"vcpu":       fmt.Sprintf("%d", runtime.NumCPU()),
+				"ds_engine":  datastoreEngine,
+			},
+		),
+		objectDefsDesc: prometheus.NewDesc(
+			prometheus.BuildFQName("spicedb", "telemetry", "object_definitions_total"),
+			"Count of the number of objects defined by the schema.",
+			nil,
+			prometheus.Labels{
+				"cluster_id": clusterID,
+				"node_id":    nodeID,
+			},
+		),
+		relationshipsDesc: prometheus.NewDesc(
+			prometheus.BuildFQName("spicedb", "telemetry", "relationships_estimate_total"),
+			"Count of the estimated number of stored relationships.",
+			nil,
+			prometheus.Labels{
+				"cluster_id": clusterID,
+				"node_id":    nodeID,
+			},
+		),
+		dispatchedDesc: prometheus.NewDesc(
+			prometheus.BuildFQName("spicedb", "telemetry", "dispatches"),
+			"Histogram of cluster dispatches performed by the instance.",
+			dynamicDispatchHistLabels,
+			prometheus.Labels{
+				"cluster_id": clusterID,
+				"node_id":    nodeID,
+			},
+		),
+	}); err != nil {
+		return fmt.Errorf("unable to register telemetry collector: %w", err)
+	}
+
+	return nil
+}
+
+type collector struct {
+	ds                datastore.Datastore
+	infoDesc          *prometheus.Desc
+	objectDefsDesc    *prometheus.Desc
+	relationshipsDesc *prometheus.Desc
+	dispatchedDesc    *prometheus.Desc
+}
+
```

Comment:
I assume this is satisfying some prometheus interface, maybe add an interface assertion?